cudax/include/cuda/experimental/__stf/internal/scalar_interface.cuh

```diff
@@ -0,0 +1,275 @@
+    }
+
+    if (dst_memory_node == data_place::host)
+    {
+      kind = cudaMemcpyDeviceToHost;
+    }
+
+    const scalar<T>& src_instance = this->instance(src_instance_id);
+    const scalar<T>& dst_instance = this->instance(dst_instance_id);
+
+    size_t sz = sizeof(T);
+
+    cuda_safe_call(cudaMemcpyAsync((void*) dst_instance.addr, (void*) src_instance.addr, sz, kind, stream));
+  }
+
+  void stream_data_allocate(
+    backend_ctx_untyped&,
+    const data_place& memory_node,
+    instance_id_t instance_id,
+    ::std::ptrdiff_t& s,
+    void**,
+    cudaStream_t stream) override
+  {
+    scalar<T>& instance = this->instance(instance_id);
+    T* base_ptr;
+
+    if (memory_node == data_place::host)
+    {
+      // Fallback to a synchronous method as there is no asynchronous host allocation API
+      cuda_safe_call(cudaStreamSynchronize(stream));
+      cuda_safe_call(cudaHostAlloc(&base_ptr, sizeof(T), cudaHostAllocMapped));
+    }
+    else
+    {
+      cuda_safe_call(cudaMallocAsync(&base_ptr, sizeof(T), stream));
+    }
+
+    instance.addr = base_ptr;
+
+    s = sizeof(T);
+  }
+
+  /// Pretend we deallocate an instance (no-op)
+  void stream_data_deallocate(backend_ctx_untyped&, const data_place&, instance_id_t, void*, cudaStream_t) override {}
+
+  bool pin_host_memory(instance_id_t) override
+  {
+    // no-op
+    return false;
+  }
+
+  void unpin_host_memory(instance_id_t) override {}
+};
+
+/**
+ * @brief Define how the CUDA stream backend must manipulate this scalar<T> interface
+ *
+ * Note that we specialize cuda::experimental::stf::shape_of to avoid ambiguous specialization
+ *
+ * @extends streamed_interface_of
+ */
+template <typename T>
+struct streamed_interface_of<scalar<T>>
+{
+  using type = scalar_stream_interface<T>;
+};
+
+/**
+ * @brief Data interface to manipulate the void interface in the CUDA graph backend
+ */
+template <typename T>
+class scalar_graph_interface : public graph_data_interface<scalar<T>>
+{
+public:
+  /// @brief Alias for the base class
+  using base = graph_data_interface<scalar<T>>;
+  /// @brief Alias for the shape type
+  using typename base::shape_t;
+
+  scalar_graph_interface(scalar<T> val)
+      : base(mv(val))
+  {}
+  scalar_graph_interface(shape_of<scalar<T>> s)
+      : base(mv(s))
+  {}
+
+  void data_allocate(
+    backend_ctx_untyped&,
+    block_allocator_untyped&,
+    const data_place&,
+    instance_id_t,
+    ::std::ptrdiff_t& s,
+    void**,
+    event_list&) override
+  {
+    s = sizeof(T);
+  }
+
+  void data_deallocate(
+    backend_ctx_untyped&, block_allocator_untyped&, const data_place&, instance_id_t, void*, event_list&) final
+  {}
+
+  cudaGraphNode_t graph_data_copy(
+    cudaMemcpyKind,
+    instance_id_t,
+    instance_id_t,
+    cudaGraph_t graph,
+    const cudaGraphNode_t* input_nodes,
+    size_t input_cnt) override
+  {
+    cudaGraphNode_t dummy;
+    cuda_safe_call(cudaGraphAddEmptyNode(&dummy, graph, input_nodes, input_cnt));
```

Comment:
This is not implemented ... Perhaps we need to add helpers ?

Reply:
More generally, the use of the "context" type rather than "stream_ctx" forces us to actually generate both codes until it's resolved dynamically, there is no good way to avoid implementing these interfaces at all on graph, and fail at runtime if the graph_ctx backend was selected.

Adding a new type of data interface is not a common task, so it's acceptable to be forced to implement it for all backends, but perhaps there should be some additional tests based on "is_invocable" to know whether a backend is implemented or not, or yet another trait class to say which backends are supported for that class.